bundles/gcp-bundle/src/main/java/org/apache/gravitino/gcs/credential/GCSTokenProvider.java

```diff
@@ -0,0 +1,206 @@
+                  String.format(
+                      "resource.name.startsWith('projects/_/buckets/%s/objects/%s')",
+                      bucketName, resourcePath));
+              // add list privilege
+              resourceExpressions.add(
+                  String.format(
+                      "api.getAttribute('storage.googleapis.com/objectListPrefix', '').startsWith('%s')",
+                      resourcePath));
+              if (writeLocations.contains(location)) {
+                writeBuckets.add(bucketName);
+                resourceExpressions =
+                    writeExpressions.computeIfAbsent(bucketName, key -> new ArrayList<>());
+                // add write privilege
+                resourceExpressions.add(
+                    String.format(
+                        "resource.name.startsWith('projects/_/buckets/%s/objects/%s')",
+                        bucketName, resourcePath));
+              }
+            });
+
+    // Construct policy according to the resource expression and privilege.
+    CredentialAccessBoundary.Builder credentialAccessBoundaryBuilder =
+        CredentialAccessBoundary.newBuilder();
+    readBuckets.forEach(
+        bucket -> {
+          List<String> readConditions = readExpressions.get(bucket);
+          if (readConditions == null || readConditions.isEmpty()) {
+            return;
+          }
+          CredentialAccessBoundary.AccessBoundaryRule.Builder builder =
+              CredentialAccessBoundary.AccessBoundaryRule.newBuilder();
+          builder.setAvailableResource(toGCSBucketResource(bucket));
+          builder.setAvailabilityCondition(
+              CredentialAccessBoundary.AccessBoundaryRule.AvailabilityCondition.newBuilder()
+                  .setExpression(String.join(" || ", readConditions))
+                  .build());
+          builder.setAvailablePermissions(Arrays.asList("inRole:roles/storage.legacyObjectReader"));
+          builder.addAvailablePermission("inRole:roles/storage.objectViewer");
+          credentialAccessBoundaryBuilder.addRule(builder.build());
+        });
+
+    writeBuckets.forEach(
+        bucket -> {
+          List<String> writeConditions = writeExpressions.get(bucket);
+          if (writeConditions == null || writeConditions.isEmpty()) {
+            return;
+          }
+          CredentialAccessBoundary.AccessBoundaryRule.Builder builder =
+              CredentialAccessBoundary.AccessBoundaryRule.newBuilder();
+          builder.setAvailableResource(toGCSBucketResource(bucket));
+          builder.setAvailabilityCondition(
+              CredentialAccessBoundary.AccessBoundaryRule.AvailabilityCondition.newBuilder()
+                  .setExpression(String.join(" || ", writeConditions))
+                  .build());
+          builder.setAvailablePermissions(Arrays.asList("inRole:roles/storage.legacyBucketWriter"));
+          credentialAccessBoundaryBuilder.addRule(builder.build());
+        });
```

Comment:
These two parts are almost the same, could you extract a common method?

Reply:
updated